tsdb/head.go

```diff
@@ -1622,15 +1624,57 @@ func (h *headIndexReader) SortedLabelValues(name string) ([]string, error) {
 
 // LabelValues returns label values present in the head for the
 // specific label name that are within the time range mint to maxt.
-func (h *headIndexReader) LabelValues(name string) ([]string, error) {
+// If matchers are specified the returned result set is reduced
+// to label values of metrics matching the matchers.
+func (h *headIndexReader) LabelValues(name string, matchers ...*labels.Matcher) ([]string, error) {
 	h.head.symMtx.RLock()
```

Comment:
Ah really, sorry I guess I'm confused about what exactly is protected by this lock. In this case this should be fine? 4110d3677f8fb310e39b5883334c6e562a215e93

The function `h.head.postings.LabelValues()` seems to have its own lock-handling, so what is `h.head.symMtx` protecting which is read in this line? Is it just to read the pointer value of `h.head.postings`?

Reply:
symMtx is actually protecting the symbols in Head but also protecting the race between adding/deleting in h.postings and updating the symbols map in Head since they need to be in sync, hence you see h.postings being modified under that lock too.